### ISSUE TITLE
Unexpected behaviour when mapping objects with collections, containing duplicate items

### ISSUE THREAD

**Issue by ds1709** [AutoMapper/AutoMapper#2188]
I'm trying to create `A` to `A` object mapping with list of `B`, where list contains duplicates of `B`.

### Source/destination types
```csharp
class A { public List<B> ListOfB { get; } = new List<B>(); }
class B
{
    public Guid Id { get; } = Guid.NewGuid();
    public override string ToString() => Id.ToString();
}
```

### Mapping configuration 1
```csharp
CreateMap<A, A>().ForMember(_ => _.ListOfB, _ => _.UseDestinationValue());
CreateMap<B, B>();
```

### Expected behavior
New list items count same, as for souce list, for each item in source list (distinct duplicates) will be created new mapped object `B`, and added into target collection so many times, as it's origin object included into source collection.

### Actual behavior
Aftare mapping, target collection size is same, as source collection, but all items are unique and no duplicates.
```
source_a.ListOfB
Count = 4
    [0]: {ccaee117-d00d-4fd3-983e-13acafda5f80}
    [1]: {ccaee117-d00d-4fd3-983e-13acafda5f80}
    [2]: {1ddb5021-ffdc-4065-8b9a-99eb65b664ea}
    [3]: {1ddb5021-ffdc-4065-8b9a-99eb65b664ea}

target_a.ListOfB
Count = 4
    [0]: {262fd3a0-257c-4c9e-b5b8-97d8678ad35d}
    [1]: {9a9ff9d1-bc19-431e-9eb4-3e3bd85fae1c}
    [2]: {a241323a-a4c5-433a-a114-e7d8873a41e1}
    [3]: {9745f625-6767-4a20-81e9-87c58765d983}
```
In source list items [0] and [1] are same, as wel as items [2] and [3], but in target list all objects are unique.

### Mapping configuration 2
```csharp
CreateMap<A, A>().ForMember(_ => _.ListOfB, _ => _.UseDestinationValue());
CreateMap<B, B>().PreserveReferences(); // Added PreserveReferences().
```

### New behaviour
Behaviour was changed. Now list maps as expected.
```
sourceListOfA
Count = 4
    [0]: {7a6d9a17-0fb4-41cc-b752-6394a8ac5e7b}
    [1]: {7a6d9a17-0fb4-41cc-b752-6394a8ac5e7b}
    [2]: {bc97de21-81bd-4702-9131-19f63e978cdd}
    [3]: {bc97de21-81bd-4702-9131-19f63e978cdd}

targetListOfA
Count = 4
    [0]: {dc4dc7d3-bb69-4ae8-8b41-128acc1f49e4}
    [1]: {dc4dc7d3-bb69-4ae8-8b41-128acc1f49e4}
    [2]: {6b6392c7-4f90-4ecb-a057-7e4671400684}
    [3]: {6b6392c7-4f90-4ecb-a057-7e4671400684}
```
In target collection items [0] and [1] are same, and items [2] and [3] are same to. But there is another problem when we add class `B1`, derived from class `B`.
```csharp
class A { public List<B> ListOfB { get; } = new List<B>(); }
class B
{
    public Guid Id { get; } = Guid.NewGuid();
    public override string ToString() => Id.ToString();
}
class B1 : B { }
```
### Mapping configuration 2
```csharp
CreateMap<A, A>().ForMember(_ => _.ListOfB, _ => _.UseDestinationValue());
CreateMap<B, B>().PreserveReferences();
CreateMap<B1, B1>().IncludeBase<B, B>();
```
### New behaviour
Now all previous configuration is broken, and mapper creates uniq objects in new list. Adding `PreserveReferences()` to `B1` mapping doesn't resolve the problem.

### Steps to reproduce
```csharp
var config = new MapperConfiguration(cfg =>
{
    cfg.CreateMap<A, A>().ForMember(_ => _.ListOfB, _ => _.UseDestinationValue());
    cfg.CreateMap<B, B>().PreserveReferences();
    cfg.CreateMap<B1, B1>().IncludeBase<B, B>();
});
config.AssertConfigurationIsValid();
var mapper = config.CreateMapper();

var source_a = new A();

var b = new B();
var b1 = new B();

source_a.ListOfB.AddRange(new[] { b, b, b1, b1 });

var target_a = mapper.Map<A>(source_a);
```

### Another scenarion (nested objects)
Object `A` contanis list of objects `B`, where one of classes, inherited from `B` (e.g. `B1`), contains reference to class `C`, and `C` has reference to one of objects `B` in root object's `A` list of `B`.
```csharp
class A { public List<B> ListOfB { get; } = new List<B>(); }
class B { }
class B1 : B { public C Item { get; set; } }
class C { public B Item { get; set; } }
```

### Steps to reproduce
```csharp
var config = new MapperConfiguration(cfg =>
{
    cfg.CreateMap<A, A>().ForMember(_ => _.ListOfB, _ => _.UseDestinationValue());
    cfg.CreateMap<B, B>().PreserveReferences();
    cfg.CreateMap<B1, B1>().IncludeBase<B, B>();
    cfg.CreateMap<C, C>().IncludeBase<B, B>();
});
config.AssertConfigurationIsValid();
var mapper = config.CreateMapper();

var source_a = new A();
            
var b = new B();
var b1 = new B1();
var c = new C { Item = b };
b1.Item = c;

source_a.ListOfB.AddRange(new[] { b, b1 });

var target_a = mapper.Map<A>(source_a);
```
```
source_a.ListOfB
Count = 2
    [0]: {MapperTestApp.Program.B}
    [1]: {MapperTestApp.Program.B1}
((B1)source_a.ListOfB[1]).Item.Item == source_a.ListOfB[0]
true

target_a.ListOfB
Count = 2
    [0]: {MapperTestApp.Program.B}
    [1]: {MapperTestApp.Program.B1}
((B1)target_a.ListOfB[1]).Item.Item == target_a.ListOfB[0]
false // expected [true]
```

**Comment by lbargaoanu** [comment]
And without self mapping?

**Comment by ds1709** [comment]
### Model
```csharp
class A { public List<B> ListOfB { get; } = new List<B>(); }
class B
{
    public Guid Id { get; } = Guid.NewGuid();
    public override string ToString() => Id.ToString();
}
class B1 : B { }
```
### Dto
```csharp
class ADto { public List<BDto> ListOfB { get; } = new List<BDto>(); }
class BDto
{
    public Guid Id { get; } = Guid.NewGuid();
    public override string ToString() => Id.ToString();
}
class B1Dto : BDto { }
```
### Call
```csharp
var config = new MapperConfiguration(cfg =>
{
    cfg.CreateMap<A, ADto>().ForMember(_ => _.ListOfB, _ => _.UseDestinationValue());
    cfg.CreateMap<B, BDto>().PreserveReferences();
    cfg.CreateMap<B1, B1Dto>().IncludeBase<B, BDto>();
});
config.AssertConfigurationIsValid();
var mapper = config.CreateMapper();

var source_a = new A();
var b = new B();
var b1 = new B1();

source_a.ListOfB.AddRange(new B[] { b, b, b1, b1 });

var target_a = mapper.Map<ADto>(source_a);
```
### Result
```
source_a.ListOfB
Count = 4
    [0]: {3238b826-1b92-46e8-987a-a369a6dd7f1f}
    [1]: {3238b826-1b92-46e8-987a-a369a6dd7f1f}
    [2]: {4bc70a15-5a82-40a6-8a2b-12483c489191}
    [3]: {4bc70a15-5a82-40a6-8a2b-12483c489191}
target_a.ListOfB
Count = 4
    [0]: {cd2d0a57-fc06-48ff-9f4c-d0ddec6e35cc}
    [1]: {83abf89e-cf63-4aaa-a9ae-45e3df87f59b}
    [2]: {a1b73e8a-e4ea-405d-9c86-42c3957b066d}
    [3]: {36c6acc1-1c05-4795-806b-c5ad3103a738}
```
So, it's same for non-self mapping.

**Comment by TylerCarlson1** [comment]
Yea each item will map to a new item.  It has not concept of duplication as far as mapping a collection is concerned.  We would either have to update collection mapper or you would have to account for that in your own.

**Comment by lbargaoanu** [comment]
Try the [MyGet](https://github.com/AutoMapper/AutoMapper/wiki/The-MyGet-Build) build. It works for me without PreserveReferences.

**Comment by lbargaoanu** [comment]
OK, that's because I made the collection writable so I can compile in LINQPad. But you have a workaround :)

**Comment by ds1709** [comment]
It still doesn't work for me. I add setter for `ListOfB` property, but it still create unique objects in target collection. ```
class ADto { public List<BDto> ListOfB { get; set; } = new List<BDto>(); }
```
Updating to version 62.0-ci-01 have no any effect.
Did you created map for derived class and include it into base class maping?

**Comment by lbargaoanu** [comment]
https://gist.github.com/lbargaoanu/8331afb734df500249f089fccad26fff

**Comment by lbargaoanu** [comment]
I guess it should work with 6.1.1 too, I just always use the latest.

**Comment by ds1709** [comment]
What is the `.Dump()` call here `var target_a = mapper.Map<ADto>(source_a).Dump();`? I cant't compile code frome your example.

**Comment by lbargaoanu** [comment]
LinqPad stuff. Remove it.

**Comment by ds1709** [comment]
Yes, it works now, but in your example you are not just add the setter on `ListOfB` property, you modified class signature and make it as field from property. Maybe thats the point, and problem in property mapper. Also, i can't just add setter in my code, cos it will break business logic in my real project.

**Comment by lbargaoanu** [comment]
That's just for quick testing, it can be a property. A private setter would work too. A fix, and a next release could be months away.

**Comment by lbargaoanu** [comment]
My gist is misleading because if you can set the id, it doesn't reflect object identity anymore :) The fix is to set PreserveReferences on all the maps, assuming that's what you want, object identity, not the same Id-s. There is no bug here. PreserveReferences is set automatically to avoid SO, but that's not the case here.

**Comment by lbargaoanu** [comment]
@ds1709 You should read [the upgrade guide](https://github.com/AutoMapper/AutoMapper/wiki/5.0-Upgrade-Guide).

**Comment by ds1709** [comment]
I set `PreserveReferences` for `A` to `ADto` and `B` to `BDto` configuration, and now it works :+1:. Thanx for the solution. But I realy can't understand some moments:
1. Why that's not the case here, why PreserveReferences doesn't set automatically? Can you provide a simple scenario where it will work?
2. Mapping works normally when I add `PreserveReferences` for `B` to `BDto` before creating map for `B1` to `B1Dto`. Why creating mapping for derived classes break `PreserveReferences` and make necessary add `PreserveReferences` for `A` to `ADto` map?
3. Why I need to add `PreserveReferences` for `A` to `ADto` map, regardless `B` has no reverse reference on `A` (same for `Dto`)? It's just a collection, containing same object twice. There's no any circular references :).
The mechanism of resolving circular references is hard to understand. Maybe it shoud be reviewed and recycled?

**Comment by lbargaoanu** [comment]
PreserveReferences does just that, but so it happens it is useful for circular references. Check [the execution plan](https://github.com/AutoMapper/AutoMapper/wiki/Understanding-your-mapping). Try setting smth and see what changes. All maps need to share the same context where the references are stored. If you break the chain, the mechanism breaks. You can find examples in [our tests](https://github.com/AutoMapper/AutoMapper/search?utf8=%E2%9C%93&q=without&type=). The mechanism works by default for circular references, but your case is different, hence the problems.

**Comment by ds1709** [comment]
In continuation of topic, regardless it closed :) , I make a little reseach and found, what after adding `PreserveReferences` for base class `B`, but before adding mapping for derived class `B1`, in execution plan appears expression blok with `ResolutionContext` and `$cachedDestination`. After adding mapping for derived class `B1` this expression block `disappear`. Adding `PreserveReferences` for `B1` mapping has no effect, result execution plan still same, as without `PreserveReferences`, and this looks strange. Loks like `PreserveReferences` just "supressed". Only after adding `PreserveReferences` for EVERY mapping (`A`, `B` and `B1`) appears expression blok with `ResolutionContext` and `$cachedDestination` (but in this case, execution plan is totaly different). Is this behaviour "by design"? If yes, maybe `PreserveReferences` shouldn't be option for exact mapping, and should be as option for a whole configuration, or adding `PreserveReferences` for type, e.g. `T`, shoud cache all object of this type during mapping process and check cache for all properties and fields of type `T` (including collections of `T`) to use cached objects. Or, maybe, it should be option for members (`ForMember<T>(m, e)`), instructing mapper to check cache only fo this member. Now it looks like I create mapping with `PreserveReferences` for an exact type, but to make it work I need add this option everywhere, and it looks inconsistent.

**Comment by ds1709** [comment]
And why `PreserveReferences` option doesn't inherits for derived class when creating mappin with `Include<,>` or `IncludeBase<,>`?

**Comment by ds1709** [comment]
I think, you should open back this issue, not like a bug, because functionality is raw and should be reworked. If I'll have free time to figure out more detailed in this question, I can provide you some unit test for your consideration.

**Comment by lbargaoanu** [comment]
Everything works as designed. The only thing I can take from this is the inheritance thing, and that's arguable. It's easy to set PreserveReferences (or anything else) globally, but that will hurt performance.
